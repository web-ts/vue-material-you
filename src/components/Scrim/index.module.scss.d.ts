declare const classNames: {
  readonly scrim: "scrim";
};
export default classNames;
export type ClassNames = "scrim";
