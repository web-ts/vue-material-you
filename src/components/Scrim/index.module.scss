.scrim {
  align-items: center;
  background-color: rgba($color: #000000, $alpha: 0.5);
  display: flex;
  flex-direction: column;
  justify-content: center;
  left: 0;
  position: fixed;
  top: 0;
  width: 100vw;
}
