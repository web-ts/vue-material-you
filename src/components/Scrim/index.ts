import useWindowDimensions from "@/composables/use-window-dimensions";
import emit from "@/utilities/emit";
import prop from "@/utilities/prop";
import { StyleValue, Teleport, Transition } from "vue";
import scss from "./index.module.scss";

export default defineComponent({
  name: "Scrim",
  props: {
    open: prop.boolean(false),
    zIndex: prop.generic(100),
    style: prop.generic<StyleValue>(),
    id: prop.generic<string>(),
    isLast: prop.boolean(false)
  },
  emits: {
    "update:modelValue": emit<boolean>(),
    open: emit()
  },
  setup(props, { emit, slots, attrs }) {
    async function onBeforeEnter() {
      await nextTick();
      emit("open");
    }

    const { height } = useWindowDimensions();

    const style = computed(() => [
      props.style,
      {
        top: 0,
        left: 0,
        position: "fixed",
        width: "100vw",
        height: `${height.value}px`,
        zIndex: props.zIndex
      }
    ]);

    const scrimProps = computed(() => ({
      ...attrs,
      id: props.id,
      style: style.value,
      tabindex: !props.isLast ? -1 : undefined,
      role: "dialog",
      "aria-hidden": !props.isLast,
      "aria-modal": "true",
      "aria-labelledby": `${props.id}_title`,
      "aria-describedby": `${props.id}_description`
    }));

    return () =>
      h(
        Teleport,
        { to: "body" },
        // Scrim Transition
        h(
          Transition,
          {
            name: "vmu-scrim",
            onBeforeEnter
          },
          () => props.open && h("div", { class: scss.scrim, ...scrimProps.value }, slots.default && slots.default())
        )
      );
  }
});
