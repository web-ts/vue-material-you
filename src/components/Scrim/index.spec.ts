import Scrim from ".";
import { describe, expect, it } from "vitest";
import { mount } from "@vue/test-utils";

describe("scrim", () => {
  it("should emit the open event", async () => {
    mount(Scrim, { attachTo: document.body, props: { open: true } });
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
