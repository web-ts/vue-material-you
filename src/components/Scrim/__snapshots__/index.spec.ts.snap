// Vitest Snapshot v1

exports[`scrim > should emit the open event 1`] = `"<div data-v-app=\\"\\"><!--teleport start--><!--teleport end--></div><transition-stub name=\\"vmu-scrim\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><div class=\\"_scrim_717d7a\\" style=\\"top: 0px; left: 0px; position: fixed; width: 100vw; height: 0px; z-index: 100;\\" tabindex=\\"-1\\" role=\\"dialog\\" aria-hidden=\\"true\\" aria-modal=\\"true\\" aria-labelledby=\\"undefined_title\\" aria-describedby=\\"undefined_description\\"></div></transition-stub>"`;
